File: app/src/main/java/net/xzos/UpgradeAll/server/log/LogDataProxy.java
package net.xzos.UpgradeAll.server.log;

import androidx.annotation.NonNull;
import androidx.lifecycle.LiveData;

import net.xzos.UpgradeAll.database.RepoDatabase;

import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;
import org.litepal.LitePal;

import java.util.ArrayList;
import java.util.Iterator;
import java.util.List;

public class LogDataProxy {
    private LogUtil Log;
    private JSONObject logJSONObject;
    private LogLiveData logLiveData;

    public LogDataProxy(@NonNull LogUtil logUtil) {
        Log = logUtil;
        logJSONObject = logUtil.getLogJSONObject();
        logLiveData = logUtil.getLogLiveData();
    }

    public List<String> getLogSort() {
        ArrayList<String> logSortList = new ArrayList<>();
        Iterator it = logJSONObject.keys();
        while (it.hasNext()) {
            logSortList.add((String) it.next());
        }
        return logSortList;
    }

    public List<String> getLogObjectId(String logSort) {
        ArrayList<String> logObjectId = new ArrayList<>();
        JSONObject logSortJson;
        try {
            logSortJson = logJSONObject.getJSONObject(logSort);
        } catch (JSONException e) {
            logSortJson = new JSONObject();
        }
        Iterator it = logSortJson.keys();
        while (it.hasNext()) {
            logObjectId.add((String) it.next());
        }
        return logObjectId;
    }

    List<String> getLogMessageList(String[] logObjectTag) throws JSONException {
        ArrayList<String> logMessageArray = new ArrayList<>();
        String logSortString = logObjectTag[0];
        String logObjectId = logObjectTag[1];
        JSONArray logMessageJSONArray = logJSONObject.getJSONObject(logSortString).getJSONArray(logObjectId);
        int len = logMessageJSONArray.length();
        for (int i = 0; i < len; i++) {
            logMessageArray.add(logMessageJSONArray.get(i).toString());
        }
        return logMessageArray;
    }

    public LiveData<List<String>> getLogMessageListLiveData(String[] logObjectTag) {
        return logLiveData.getLogLiveDataList(logObjectTag);
    }

    public void clearLogAll() {
        logJSONObject = new JSONObject();
        Log.setLogJSONObject(logJSONObject);
        logLiveData.setLogJSONObject(logJSONObject);
    }

    public void clearLogSort(String logSort) {
        logJSONObject.remove(logSort);
        Log.setLogJSONObject(logJSONObject);
        logLiveData.setLogJSONObject(logJSONObject);
    }

    public void clearLogMessage(String[] logObjectTag) throws JSONException {
        String logSortString = logObjectTag[0];
        String logObjectId = logObjectTag[1];
        logJSONObject.getJSONObject(logSortString).remove(logObjectId);
        Log.setLogJSONObject(logJSONObject);
        logLiveData.setLogJSONObject(logJSONObject);
    }

    public String getLogAllToString() {
        List<String> sortList = getLogSort();
        StringBuilder fullLogString = new StringBuilder();
        for (String logSort : sortList) {
            fullLogString.append(getLogStringBySort(logSort));
        }
        return fullLogString.toString();
    }

    public String getLogStringBySort(String logSort) {
        StringBuilder fullLogString = new StringBuilder(logSort + "\n");
        List<String> sortDatabaseIdList = getLogObjectId(logSort);
        for (String databaseIdString : sortDatabaseIdList) {
            String logString = getLogMessageToString(new String[]{logSort, databaseIdString});
            if (logString != null)
                logString = logString.split("\n", 2)[1];
            else logString = "";
            fullLogString.append(logString).append("\n");
        }
        return fullLogString.toString();
    }

    public String getLogMessageToString(String[] logObjectTag) {
        String sort = logObjectTag[0];
        String name = "    " + getNameFromId(logObjectTag[1]);
        StringBuilder logMessageString = new StringBuilder();
        List<String> logMessageArray;
        try {
            logMessageArray = getLogMessageList(logObjectTag);
        } catch (JSONException e) {
            return null;
        }
        for (String logMessage : logMessageArray)
            logMessageString.append("        ").append(logMessage).append("\n");
        return sort + "\n" + name + "\n" + logMessageString;
    }

    public static String getNameFromId(@NonNull String databaseIdString) {
        String name;
        try {
            RepoDatabase repoDatabase = LitePal.find(RepoDatabase.class, Integer.parseInt(databaseIdString));
            name = repoDatabase.getName();
        } catch (Throwable e) {
            name = databaseIdString;
        }
        return name;
    }
}

File: app/src/main/kotlin/net/xzos/upgradeAll/server/log/LogDataProxy.kt
package net.xzos.upgradeAll.server.log

import androidx.lifecycle.LiveData
import net.xzos.upgradeAll.data.database.manager.AppDatabaseManager
import net.xzos.upgradeAll.data.json.nongson.ObjectTag
import org.json.JSONException

internal class LogDataProxy(private val logMap: MutableMap<String, MutableMap<String, MutableList<String>>>) {
    private val logLiveData = LogUtil.logLiveData

    internal val logSort: MutableSet<String>
        get() = logMap.keys

    val liveDataLogSortList: LiveData<MutableSet<String>>
        get() = logLiveData.sortList

    val logAllToString: String
        get() {
            val sortList = logSort
            val fullLogString = StringBuilder()
            for (logSort in sortList) {
                fullLogString.append(getLogStringBySort(logSort))
            }
            return fullLogString.toString()
        }

    internal fun getLogObjectId(logSort: String): MutableSet<String> =
            this.logMap[logSort]?.keys ?: mutableSetOf()

    internal fun getLogMessageList(objectTag: ObjectTag): List<String> {
        val logSort = objectTag.sort
        val logObjectId = objectTag.id
        return logMap[logSort]?.get(logObjectId) ?: listOf()
    }

    fun getLiveDataLogObjectIdList(logSort: String): LiveData<MutableSet<String>> {
        return logLiveData.getIdListInSort(logSort)
    }

    fun getLogMessageListLiveData(objectTag: ObjectTag): LiveData<List<String>> {
        return logLiveData.getLogMassageList(objectTag)
    }

    fun clearLogAll() {
        logMap.clear()
        LogUtil.notifyObserver()
    }

    fun clearLogSort(logSort: String) {
        logMap.remove(logSort)
        LogUtil.notifyObserver()
    }

    fun clearLogMessage(logObjectTag: Pair<String, String>) {
        val logSortString = logObjectTag.first
        val logObjectId = logObjectTag.second
        this.logMap[logSortString]?.remove(logObjectId)
        LogUtil.notifyObserver()
    }

    fun getLogStringBySort(logSort: String): String {
        val fullLogString = StringBuilder(logSort + "\n")
        val sortDatabaseIdList = getLogObjectId(logSort)
        for (databaseIdString in sortDatabaseIdList) {
            var logString = getLogMessageToString(ObjectTag(logSort, databaseIdString))
            logString = if (logString != null)
                logString.split("\n".toRegex(), 2).toTypedArray()[1]
            else
                ""
            fullLogString.append(logString).append("\n")
        }
        return fullLogString.toString()
    }

    private fun getLogMessageToString(logObjectTag: ObjectTag): String? {
        val sort = logObjectTag.sort
        val name = "    " + getNameFromId(logObjectTag.id)
        val logMessageString = StringBuilder()
        val logMessageArray: List<String>
        try {
            logMessageArray = getLogMessageList(logObjectTag)
        } catch (e: JSONException) {
            return null
        }

        for (logMessage in logMessageArray)
            logMessageString.append("        ").append(logMessage).append("\n")
        return sort + "\n" + name + "\n" + logMessageString
    }

    internal fun getNameFromId(databaseIdString: String): String? = try {
        AppDatabaseManager.getDatabase(databaseId = databaseIdString.toInt()).name
    } catch (e: Throwable) {
        databaseIdString
    }
}